### PR TITLE
Add support for salted encryption key

### PR DESCRIPTION
**Pull request by ChrisSchinnerl**
Closes https://github.com/SiaFoundation/renterd/issues/1559

To distinguish between the different versions of encryption keys, they are prefixed with a single byte in the database.
I'm also open to suggestions when it comes to naming. "salted keys" is more of a working title.

### CODE REVIEW

File: internal/utils/keys.go [hunk 1]
```diff
@@ -19,6 +20,12 @@ func (key *MasterKey) DeriveAccountsKey(workerID string) AccountsKey {
 	return AccountsKey(key.deriveSubKey(keyPath))
 }
 
+// DeriveAccountsKey derives an accounts key from a masterkey which is used
```
Comment (peterjan) [inline internal/utils/keys.go:23]: copy pasta

File: internal/utils/keys.go [hunk 2]
```diff
@@ -30,6 +37,13 @@ func (key *MasterKey) DeriveContractKey(hostKey types.PublicKey) types.PrivateKe
 	return pk
 }
 
+func (key *UploadKey) DeriveKey(salt *[32]byte) [32]byte {
```
Comment (peterjan) [inline internal/utils/keys.go:40]: `nit` could use a docstring

File: object/encryptionkey.go [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+type EncryptionKeyType int
+
+const (
+	EncryptionKeyTypeBasic = EncryptionKeyType(iota + 1)
+	EncryptionKeyTypeSalted
+)
+
+// A EncryptionKey can encrypt and decrypt messages.
+type EncryptionKey struct {
+	entropy *[32]byte         `json:"-"`
```
Comment (peterjan) [inline object/encryptionkey.go:41]: both fields are not exposed so they're not json encoded by default, making the annotation redundant?

File: object/encryptionkey.go [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+	}
+	return nil
+}
+
+type EncryptionOptions struct {
+	Offset uint64
+	Key    *utils.UploadKey
+}
+
+var ErrKeyType = errors.New("invalid key type")
```
Comment (peterjan) [inline object/encryptionkey.go:147]: `nit` can we move these up so we have `const` then `var` then `types` and so on

Comment (peterjan) [reply object/encryptionkey.go:147]: (I can see why you define the type and then the methods though, but consts and vars should go up top imo)

**Review by peterjan — APPROVED** [review]
I thought about a better name but couldn't come up with a better name than you did. If it's hard to find a better name, being descriptive is likely the best option so perhaps `SaltedEncryptionKey` isn't the worst thing. For a second I was thinking instead of finding a name for the new type of key we could find a name for the old but I guess we can't do that because of the existing `key` prefix...

**Comment by ChrisSchinnerl** [comment]
@peterjan Sry hit update by rebase instead of merge commit but the changes are all in the latest commit.

**Review by peterjan — APPROVED** [review]

**Pull request merged**